hal/src/sercom/v2/uart/reg.rs

```diff
@@ -0,0 +1,411 @@
+    /// overflow occurs. Otherwise, it will not be issued until its place within
+    /// the data stream.
+    #[inline]
+    pub(super) fn set_immediate_overflow_notification(&mut self, set: bool) {
+        self.usart().ctrla.modify(|_, w| w.ibon().bit(set));
+    }
+
+    /// Get the current immediate overflow notification setting
+    #[inline]
+    pub(super) fn get_immediate_overflow_notification(&self) -> bool {
+        self.usart().ctrla.read().ibon().bit()
+    }
+
+    /// Run in standby mode
+    ///
+    /// When set, the UART peripheral will run in standby mode. See the
+    /// datasheet for more details.
+    #[inline]
+    pub(super) fn set_run_in_standby(&mut self, set: bool) {
+        self.usart().ctrla.modify(|_, w| w.runstdby().bit(set));
+    }
+
+    /// Get the current run in standby mode
+    #[inline]
+    pub(super) fn get_run_in_standby(&self) -> bool {
+        self.usart().ctrla.read().runstdby().bit()
+    }
+
+    /// Enable or disable IrDA encoding. The pulse length controls the minimum
+    // pulse length that is required for a pulse to be accepted by the IrDA
+    /// receiver with regards to the serial engine clock period.
+    /// See datasheet for more information.
+    #[inline]
+    pub(super) fn set_irda_encoding(&mut self, pulse_length: Option<u8>) {
+        match pulse_length {
+            Some(l) => {
+                self.usart().rxpl.write(|w| unsafe { w.rxpl().bits(l) });
+                self.usart().ctrlb.modify(|_, w| w.enc().bit(true));
+            }
+            None => {
+                self.usart().ctrlb.modify(|_, w| w.enc().bit(false));
+            }
+        }
+    }
+
+    /// Get the current IrDA encoding setting. The return type is the pulse
+    /// length wrapped in an [`Option`].
+    #[inline]
+    pub(super) fn get_irda_encoding(&self) -> Option<u8> {
+        if !self.usart().ctrlb.read().enc().bit() {
+            return None;
+        }
+
+        Some(self.usart().rxpl.read().bits())
+    }
+
+    /// Clear specified interrupt flags
+    #[inline]
+    pub fn clear_flags(&mut self, bits: u8) {
+        self.usart().intflag.modify(|_, w| unsafe { w.bits(bits) });
+    }
+
+    /// Read interrupt flags
+    #[inline]
+    pub fn read_flags(&self) -> u8 {
+        self.usart().intflag.read().bits()
+    }
+
+    /// Enable specified interrupts
+    #[inline]
+    pub fn enable_interrupts(&mut self, bits: u8) {
+        self.usart().intenset.write(|w| unsafe { w.bits(bits) });
+    }
+
+    /// Disable specified interrupts
+    #[inline]
+    pub fn disable_interrupts(&mut self, bits: u8) {
+        self.usart().intenclr.write(|w| unsafe { w.bits(bits) });
+    }
+
+    /// Clear specified status flags
+    #[inline]
+    pub fn clear_status(&mut self, bits: u16) {
+        self.usart().status.modify(|_, w| unsafe { w.bits(bits) });
+    }
+
+    /// Read status flags
+    #[inline]
+    pub fn read_status(&self) -> u16 {
+        self.usart().status.read().bits()
+    }
```

Comment:
I don't think I follow my own advice here, but I just realized that it might be better for these functions to take `Flags` and `Status` types, rather than raw bits.

Right now, the `unsafe` code has to rely on the caller to provide valid values, which means it could technically be unsound if a HAL author made a mistake. But if you used the `Flags` and `Status` types, it would make the check entirely local to the function and truly safe.

I think it would clean up some of the downstream code too.